File: Leetcode/121-Best-Time-to-Buy-and-Sell-Stock/121-Best-Time-to-Buy-and-Sell-Stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        maxProfit = 0
        minCurrentBuy = prices[0] + 1

        for price in prices:
            if price < minCurrentBuy:
                minCurrentBuy = price
            else:
                currentprofit = price - minCurrentBuy
                if currentprofit > maxProfit:
                    maxProfit = currentprofit

        return maxProfit

File: Leetcode/121-Best-Time-to-Buy-and-Sell-Stock/121-Best-Time-to-Buy-and-Sell-Stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int pricesLength = prices.length, bestSellPrice = prices[pricesLength - 1];
        int maxProfit = 0;

        for (int i = pricesLength - 2; i >= 0; i--) {
            maxProfit = Math.max(maxProfit, bestSellPrice - prices[i]);
            bestSellPrice = Math.max(bestSellPrice, prices[i]);
        }

        return maxProfit;
    }
}